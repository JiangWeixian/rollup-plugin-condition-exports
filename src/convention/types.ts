import type { PackageContext } from './context'
import type { Awaitable } from '@antfu/utils'

export type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>

export interface PageOptions {
  /**
   * Paths to the directory to search for page components.
   * @default 'src/exports'
   */
  dir: string
  /**
   * @example if dir is `src/exports` base will be `src`
   * @example if dir is `components` base will be `components`
   */
  base: string
}

export interface PageResolver {
  resolveExtensions: () => string[]
  resolvePkg: (ctx: PackageContext) => Awaitable<any>
}

/**
 * Plugin options.
 */
interface Options {
  /**
   * Paths to the directory to search for page components.
   * @default 'src/exports'
   */
  dirs: string | PageOptions
  /**
   * Paths to the directory output bundle files
   * @default 'dist'
   */
  outDir: string
  /**
   * Paths to the directory output dts files
   * @default 'dist'
   */
  declarationDir: string
  /**
   * Valid file extensions
   * @default ['tsx', 'ts', 'jsx', 'jsx']
   */
  extensions: string[]
  /**
   * CJS format extension
   * @default 'cjs'
   */
  cjsExtension: string
  /**
   * ESM format extension
   * @default 'mjs'
   */
  esmExtension: string
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[]
  /**
   * List of exports condition names
   * @default ['node', 'browser', 'default']
   */
  conditions: string[]
  /**
   * Custom generated routes
   */
  onRoutesGenerated?: (routes: any[]) => Awaitable<any[] | void>
}

export type UserOptions = Partial<Options>

export interface ResolvedOptions
  extends Omit<
    Options,
    'pagesDir' | 'replaceSquareBrackets' | 'nuxtStyle' | 'syncIndex' | 'moduleId' | 'dirs'
  > {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string
  /**
   * Resolved page dirs
   */
  dirs: PageOptions
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp
}
