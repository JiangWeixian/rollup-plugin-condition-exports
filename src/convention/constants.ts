export const dynamicRouteRE = /^\[(.+)\]$/
export const cacheAllRouteRE = /^\[\.{3}(.*)\]$/
export const replaceDynamicRouteRE = /^\[(?:\.{3})?(.*)\]$/

export const countSlashRE = /\//g

export const replaceIndexRE = /\/?index$/

export const defaultExtensions = ['tsx', 'jsx', 'ts', 'js']
