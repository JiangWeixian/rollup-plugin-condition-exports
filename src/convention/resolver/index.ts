/* eslint-disable import/no-extraneous-dependencies */
import { join } from 'node:path'

import { isEmpty } from 'lodash-es'

import {
  buildReactRoutePath,
  countSlash,
  debug,
} from '../utils'

import type { PackageContext } from '../context'
import type {
  Optional,
  PageResolver,
  ResolvedOptions,
} from '../types'

export interface ReactRouteBase {
  children?: ReactRouteBase[]
  leaf: boolean
  element?: string
  index?: boolean
  path?: string
  condition?: string
  rawRoute: string
}

export interface ReactRoute extends Omit<Optional<ReactRouteBase, 'path'>, 'children'> {
  children?: ReactRoute[]
}

function prepareRoutes(routes: ReactRoute[], options: ResolvedOptions, parent?: ReactRoute) {
  for (const route of routes) {
    if (parent) {
      route.path = route.path?.replace(/^\//, '')
    }

    if (route.children) {
      route.children = prepareRoutes(route.children, options, route)
    }
  }

  return routes
}

async function computeExports(ctx: PackageContext): Promise<ReactRoute[]> {
  const pageRoutes = [...ctx.pageRouteMap.values()].sort(
    (a, b) => countSlash(a.route) - countSlash(b.route),
  )

  const routes: ReactRouteBase[] = []

  pageRoutes.forEach((page) => {
    const pathNodes = page.route.split('/')
    const element = page.path
    let parentRoutes = routes

    for (let i = 0; i < pathNodes.length; i++) {
      const node = pathNodes[i]

      const route: ReactRouteBase = {
        path: '',
        leaf: i === pathNodes.length - 1,
        rawRoute: pathNodes.slice(0, i + 1).join('/'),
      }

      if (i === pathNodes.length - 1) {
        route.element = element
        route.condition = ctx.options.conditions.includes(node) ? node : undefined
      }

      const isIndexRoute = node.endsWith('index')

      if (!route.path && isIndexRoute) {
        route.path = ''
      } else if (!isIndexRoute) {
        route.path = buildReactRoutePath(node)
      }

      // Check parent exits
      const parent = parentRoutes.find((parent) => {
        return pathNodes.slice(0, i).join('/') === parent.rawRoute
      })

      // only enable condition exports on dirs
      if (parent) {
        route.condition = parent.element ? undefined : route.condition
      }

      // only nested route on condition exports dir
      if (parent && route.condition) {
        // Make sure children exits in parent
        parent.children = parent.children || []
        // Append to parent's children
        parentRoutes = parent.children
      }

      const exits = parentRoutes.some((parent) => {
        return pathNodes.slice(0, i + 1).join('/') === parent.rawRoute
      })
      if (!exits) {
        parentRoutes.push(route)
        // absolute route path
        route.path = !route.condition ? join(parent?.path ?? '', route.path ?? '') : parent?.path
      }
    }
  })

  // sort by dynamic routes
  let finalRoutes = prepareRoutes(routes, ctx.options)

  finalRoutes = (await ctx.options.onRoutesGenerated?.(finalRoutes)) || finalRoutes

  return finalRoutes
}

const formatDeclarationPath = (element: string, ctx: PackageContext) => {
  const childDir = ctx.options.dirs.dir.replace(ctx.options.dirs.base, '')
  return `${element?.replace(ctx.options.outDir, join(ctx.options.declarationDir, childDir))}.d.ts`
}

const _resolvePkg = (routes: ReactRoute[], ctx: PackageContext, pkg: any = {}) => {
  for (const route of routes) {
    const path = route.path && route.path !== '.' ? `./${route.path}` : '.'
    if (route.leaf) {
      if (path === '.') {
        pkg.main = `${route.element}.${ctx.options.cjsExtension}`
        pkg.module = `${route.element}.${ctx.options.esmExtension}`
        pkg.types = formatDeclarationPath(route.element!, ctx)
      } else {
        pkg.typesVersions[`${path.slice(2)}`] = [formatDeclarationPath(route.element!, ctx)]
      }
      pkg.exports[`${path}`] = pkg.exports[`${path}`] ?? {}
      let subExports = pkg.exports[`${path}`]
      if (route.condition) {
        pkg.exports[`${path}`][route.condition] = pkg.exports[`${path}`][route.condition] ?? {}
        subExports = pkg.exports[`${path}`][route.condition]
      }
      subExports.import = `./${route.element}.${ctx.options.esmExtension}`
      subExports.require = `./${route.element}.${ctx.options.cjsExtension}`
      subExports.types = `./${formatDeclarationPath(route.element!, ctx)}`
    }
    if (route.children) {
      _resolvePkg(route.children, ctx, pkg)
    }
  }
}

export async function resolvePkg(ctx: PackageContext) {
  const finalRoutes = await computeExports(ctx)
  debug.search('final routes %o', finalRoutes)
  const pkg: any = {
    exports: {},
    typesVersions: {},
  }
  pkg.exports['./package.json'] = './package.json'
  _resolvePkg(finalRoutes, ctx, pkg)
  pkg.typesVersions = !isEmpty(pkg.typesVersions) ? { '*': pkg.typesVersions } : undefined
  return pkg
}

export function resolver(): PageResolver {
  return {
    resolveExtensions() {
      return ['tsx', 'jsx', 'ts', 'js']
    },
    async resolvePkg(ctx) {
      return resolvePkg(ctx)
    },
  }
}
