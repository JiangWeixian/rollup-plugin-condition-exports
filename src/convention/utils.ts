import { resolve } from 'path'
import { URLSearchParams } from 'url'
import { slash } from '@antfu/utils'
import { cacheAllRouteRE, countSlashRE, dynamicRouteRE, replaceDynamicRouteRE } from './constants'
import Debug from 'debug'

import type { ResolvedOptions } from './types'

export const debug = {
  env: Debug('rpce:env'),
  options: Debug('rpce:options'),
  pages: Debug('rpce:pages'),
  search: Debug('rpce:search'),
  pkg: Debug('rpce:pkg'),
}

export function extsToGlob(extensions: string[]) {
  return extensions.length > 1 ? `{${extensions.join(',')}}` : extensions[0] || ''
}

export function countSlash(value: string) {
  return (value.match(countSlashRE) || []).length
}

function isPagesDir(path: string, options: ResolvedOptions) {
  for (const page of options.dirs) {
    const dirPath = slash(resolve(options.root, page.dir))
    if (path.startsWith(dirPath)) return true
  }
  return false
}

export function isTarget(path: string, options: ResolvedOptions) {
  return isPagesDir(path, options) && options.extensionsRE.test(path)
}

export function isDynamicRoute(routePath: string) {
  return dynamicRouteRE.test(routePath)
}

export function isCatchAllRoute(routePath: string) {
  return cacheAllRouteRE.test(routePath)
}

export function normalizeName(name: string, isDynamic: boolean) {
  if (!isDynamic) return name

  return name.replace(replaceDynamicRouteRE, '$1')
}

export function buildReactRoutePath(node: string): string | undefined {
  const isDynamic = isDynamicRoute(node)
  const isCatchAll = isCatchAllRoute(node)
  const normalizedName = normalizeName(node, isDynamic)

  if (isDynamic) {
    if (isCatchAll) return '*'

    return `:${normalizedName}`
  }

  return `${normalizedName}`
}

export function parsePageRequest(id: string) {
  const [moduleId, rawQuery] = id.split('?', 2)
  const query = new URLSearchParams(rawQuery)
  const pageId = query.get('id')
  return {
    moduleId,
    query,
    pageId,
  }
}
