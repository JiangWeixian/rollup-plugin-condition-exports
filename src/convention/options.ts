import { resolve } from 'path'
import { slash, toArray } from '@antfu/utils'
import { getPageDirs } from './files'

import { defaultExtensions } from './constants'
import type { ImportModeResolver, ResolvedOptions, UserOptions } from './types'

function resolvePageDirs(dirs: UserOptions['dirs'], root: string, exclude: string[]) {
  dirs = toArray(dirs)
  return dirs.flatMap((dir) => {
    const option = typeof dir === 'string' ? { dir, baseRoute: '' } : dir

    option.dir = slash(resolve(root, option.dir)).replace(`${root}/`, '')
    option.baseRoute = option.baseRoute.replace(/^\//, '').replace(/\/$/, '')

    return getPageDirs(option, root, exclude)
  })
}

export const syncIndexResolver: ImportModeResolver = (filepath, options) => {
  for (const page of options.dirs) {
    if (page.baseRoute === '' && filepath.startsWith(`/${page.dir}/index`)) return 'sync'
  }
  return 'async'
}

export function resolveOptions(userOptions: UserOptions, viteRoot?: string): ResolvedOptions {
  const {
    dirs = ['src/exports'],
    exclude = ['node_modules', '.git', '**/__*__/**'],
    routeNameSeparator = '-',
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
    cjsExtension = 'cjs',
    esmExtension = 'mjs',
    outDir = 'dist',
    declarationDir = 'dist',
  } = userOptions

  const root = viteRoot || slash(process.cwd())

  const importMode = userOptions.importMode || 'async'

  const extensions = userOptions.extensions || defaultExtensions

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`)

  const resolvedDirs = resolvePageDirs(dirs, root, exclude)
  const resolvedOutDir = slash(resolve(root, outDir)).replace(`${root}/`, '')
  const resolvedDeclarationDir = slash(resolve(root, declarationDir)).replace(`${root}/`, '')
  const resolvedOptions: ResolvedOptions = {
    dirs: resolvedDirs,
    outDir: resolvedOutDir,
    declarationDir: resolvedDeclarationDir,
    root,
    extensions,
    importMode,
    exclude,
    extensionsRE,
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
    routeNameSeparator,
    cjsExtension,
    esmExtension,
  }

  return resolvedOptions
}
