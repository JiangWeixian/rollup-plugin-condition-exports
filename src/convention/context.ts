import { extname, join, resolve } from 'path'
import { slash, toArray } from '@antfu/utils'
import { resolveOptions } from './options'
import { getPageFiles } from './files'
import { debug } from './utils'
import { resolver } from './resolver'

import type { PageOptions, ResolvedOptions, UserOptions } from './types'

export interface PageRoute {
  path: string
  route: string
}

export class PageContext {
  private _pageRouteMap = new Map<string, PageRoute>()

  rawOptions: UserOptions
  root: string
  options: ResolvedOptions
  resolver = resolver()

  constructor(userOptions: UserOptions, viteRoot: string = process.cwd()) {
    this.rawOptions = userOptions
    this.root = slash(viteRoot)
    debug.env('root', this.root)
    this.options = resolveOptions(userOptions, this.root)
    debug.options(this.options)
  }

  async addPage(path: string | string[], pageDir: PageOptions) {
    debug.pages('add', path)
    for (const p of toArray(path)) {
      const pageDirPath = slash(resolve(this.root, pageDir.dir))
      const route = slash(
        join(pageDir.baseRoute, p.replace(`${pageDirPath}/`, '').replace(extname(p), '')),
      )
      this._pageRouteMap.set(p, {
        path: p,
        route,
      })
    }
  }

  async removePage(path: string) {
    debug.pages('remove', path)
    this._pageRouteMap.delete(path)
  }

  async resolveRoutes() {
    return this.resolver.resolveRoutes(this)
  }

  async searchGlob() {
    const pageDirFiles = this.options.dirs.map((page) => {
      const pagesDirPath = slash(resolve(this.options.root, page.dir))
      const files = getPageFiles(pagesDirPath, this.options)
      debug.search(page.dir, files)
      return {
        ...page,
        files: files.map((file) => slash(file)),
      }
    })

    for (const page of pageDirFiles) await this.addPage(page.files, page)
  }

  get pageRouteMap() {
    return this._pageRouteMap
  }
}
