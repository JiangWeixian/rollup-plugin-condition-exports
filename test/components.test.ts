import { describe, it, expect } from 'vitest'
import path from 'path'

import { PackageContext } from '../src/convention/context'

const FIXTURES = path.resolve(__dirname, './fixtures')

describe('basic', () => {
  it('search inputs', async () => {
    const ctx = new PackageContext(
      { dirs: 'components', exclude: ['**/**/!(index).{tsx,ts}'] },
      FIXTURES,
    )
    await ctx.searchGlob()
    const inputs = await ctx.resolveInputs()
    expect(inputs.map((v) => v.replace(FIXTURES, ''))).toMatchInlineSnapshot(`
      [
        "/components/index.ts",
        "/components/button/index.ts",
      ]
    `)
  })
  it('resolve pkg', async () => {
    const ctx = new PackageContext(
      { dirs: 'components', exclude: ['**/**/!(index).{tsx,ts}'] },
      FIXTURES,
    )
    await ctx.searchGlob()
    const pkg = await ctx.resolvePkg()
    expect(pkg).toMatchSnapshot()
  })
})
