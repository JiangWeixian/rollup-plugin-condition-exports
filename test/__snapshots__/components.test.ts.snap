// Vitest Snapshot v1

exports[`basic > resolve pkg 1`] = `
{
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts",
    },
    "./button": {
      "import": "./dist/button/index.mjs",
      "require": "./dist/button/index.cjs",
      "types": "./dist/button/index.d.ts",
    },
    "./package.json": "./package.json",
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "typesVersions": {
    "*": {
      "button": [
        "dist/button/index.d.ts",
      ],
    },
  },
}
`;
